File: Algorithm/Other/at coder/Grundy number/atcoder/B34/B34.go
// 以下は、**Go 1.20.6** に対応した `X=2, Y=3` 固定の Nim型取り石ゲームにおける
// **正確・高速・メモリ効率良好** な解法です。

// ---

// ## ✅ 解法方針

// * `X=2`, `Y=3` の場合、Grundy数は **mod 5 で周期性**がある。
// * 各山の石数 $A_i$ に対し、Grundy数 $G(A_i) = grundy[A_i \mod 5]$ を求め、XOR を取る。
// * XOR 和が `0` → **後手勝利（"Second"）**
//   XOR 和が `≠0` → **先手勝利（"First"）**

// ---

// ## ✅ Goコード（コメント付き、型注釈、処理時間・メモリ考慮）

// ```go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

// determineWinner 判定関数：取り石ゲームの勝者を返す
//
// 引数:
//   A []int64 : 各山の石の個数
//
// 戻り値:
//   string : "First" または "Second"
//
// 計算量:
//   時間計算量: O(N)
//   空間計算量: O(1)（Grundyテーブルのみ使用）
func determineWinner(A []int64) string {
	// Grundy数: X=2, Y=3 のとき周期5で以下の通り
	grundyMod := [5]int{0, 0, 1, 1, 2}

	xorSum := 0
	for _, a := range A {
		g := grundyMod[a%5]
		xorSum ^= g
	}

	if xorSum == 0 {
		return "Second"
	}
	return "First"
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Split(bufio.ScanWords)

	// readInt reads the next token and converts it to int
	readInt := func() int {
		scanner.Scan()
		v, _ := strconv.Atoi(scanner.Text())
		return v
	}

	// readInt64 reads the next token and converts it to int64
	readInt64 := func() int64 {
		scanner.Scan()
		v, _ := strconv.ParseInt(scanner.Text(), 10, 64)
		return v
	}

	N := readInt() // number of mountains
	_ = readInt()  // X (unused, fixed = 2)
	_ = readInt()  // Y (unused, fixed = 3)

	A := make([]int64, N)
	for i := 0; i < N; i++ {
		A[i] = readInt64()
	}

	result := determineWinner(A)
	fmt.Println(result)
}

// ## ✅ 処理例（標準入力）

// ```
// 2 2 3
// 5 8
// ```

// → 出力

// ```
// First
// ```

// ---

// ## ✅ Grundy数とXORの仕組み（図解）

// | A\[i] | A\[i] % 5 | Grundy\[A\[i] % 5] |
// | ----- | --------- | ------------------ |
// | 5     | 0         | 0                  |
// | 8     | 3         | 1                  |

// XOR: `0 ^ 1 = 1` → 非0 → **"First"**

// ---

// ## ✅ 時間・空間効率まとめ

// | 指標    | 内容                  |
// | ----- | ------------------- |
// | 時間計算量 | O(N)                |
// | 空間計算量 | O(1)（配列 + 定数）       |
// | 最大N   | 100,000（問題制約）       |
// | Ai最大  | 10^18（int64で安全に扱える） |

// ---

// ## ✅ 使用した言語仕様

// * `bufio.Scanner`：高速な標準入力処理
// * `strconv.ParseInt`：10^18 に対応するために `int64` を使用
// * `make([]int64, N)`：O(N) 入力配列

// ---

// ## ✅ 拡張可能な点

// * 任意の `X, Y` に対して Grundy数の `mex()` を動的に計算する汎用化
// * 動的メモ化（`map[int64]int`）で周期発見の自動化
// * テストファイルの標準入力対応

// ---

// 必要であれば：

// * テスト用入力ファイルの作成コード
// * 任意 `X`, `Y` に対応する汎用Nim解法
//   などもお手伝いできます！

File: Algorithm/Other/at coder/Grundy number/atcoder/B34/B34.py
# 以下は、**Python (CPython 3.11.4)** を用いた、`X = 2, Y = 3` 固定の **取り石ゲーム（Game 3）** に対する完全解答です。

# ---

# ## ✅ 解法概要（Nimゲーム + Grundy数周期活用）

# * 操作は `X = 2`, `Y = 3` 固定。
# * 各山の石の数に応じて Grundy 数（勝敗情報）を求める。
# * `XOR` 和が 0 → **後手必勝（"Second"）**、非 0 → **先手必勝（"First"）**。
# * `Grundy数 mod 5` が周期になることが知られているので高速化可能。

# ---

# ## ✅ Python コード（型注釈付き、処理時間・メモリ効率◎）

# ```python
import sys
from typing import List

def determine_winner(A: List[int]) -> str:
    """
    石取りゲームの勝者を判定する関数。
    
    パラメータ:
        A (List[int]) : 各山の石の個数（最大 10^18）
    
    戻り値:
        str : 'First' または 'Second'
    
    時間計算量: O(N)
    空間計算量: O(1) （定数サイズのGrundy数テーブルのみ使用）
    """
    # Grundy数の周期テーブル (X=2, Y=3 のとき mod 5 で周期)
    grundy_mod: List[int] = [0, 0, 1, 1, 2]

    xor_sum: int = 0
    for a in A:
        g = grundy_mod[a % 5]
        xor_sum ^= g

    return "Second" if xor_sum == 0 else "First"

def main() -> None:
    """
    標準入力から読み込んで結果を出力するメイン関数。
    """
    input_lines = sys.stdin.read().split()
    _: int = int(input_lines[0])
    # input_lines[1] と input_lines[2] は X, Y だが今回は固定なので無視
    A: List[int] = list(map(int, input_lines[3:]))  # A1〜AN

    result: str = determine_winner(A)
    print(result)

if __name__ == "__main__":
    main()
# ```

# ---

# ## ✅ 各処理の補足と理由

# | 処理                             | 内容                     | 備考               |
# | ------------------------------ | ---------------------- | ---------------- |
# | `grundy_mod = [0, 0, 1, 1, 2]` | Grundy数は mod 5 で周期化できる | X=2, Y=3 固定による性質 |
# | `a % 5`                        | 石数を周期化しGrundy数取得       | 高速かつ O(1)        |
# | `xor_sum ^= g`                 | 各山のGrundy数のXOR合計       | Nimゲーム理論の勝敗判定に必須 |
# | `xor_sum == 0`                 | 後手必勝の条件                | XOR=0 が負け状態を示す   |

# ---

# ## ✅ 処理時間・メモリ使用量の見積もり

# | 要素     | 使用量                                 |
# | ------ | ----------------------------------- |
# | 時間計算量  | O(N)（最大 100,000 山でも余裕）              |
# | メモリ使用量 | O(N) + O(1)（入力 + Grundy数表のみ）        |
# | 対応範囲   | Ai ≤ 10^18 対応済み（Python の int は任意精度） |

# ---

# ## ✅ 入出力例

# ### 入力

# ```
# 2 2 3
# 5 8
# ```

# ### 出力

# ```
# First
# ```

# ---

# ## ✅ テスト用の補足コード例（ローカルテスト用）

# ```python
# if __name__ == "__main__":
#     # ローカルテスト用
#     import io
#     sys.stdin = io.StringIO("2 2 3\n5 8\n")
#     main()
# ```

# ---

# ## ✅ 拡張対応も可能

# 必要に応じて：

# * 任意の `X`, `Y` に対応する汎用Grundy数列生成
# * メモ化・周期発見の自動化
# * 一般化 Nim ゲームへの拡張

# なども対応可能です。必要があればお知らせください。
